src/vm_cpu.rs

```diff
@@ -285,6 +285,79 @@ impl VmPmuState {
             )
         }
     }
+
+    fn get_counter_csr_value(counter_index: u64) -> Option<u64> {
+        use drivers::pmu::PmuInfo;
+        use riscv_regs::CSR_CYCLE;
+        let pmu_info = PmuInfo::get();
+        if let Ok(csr) = pmu_info.get_csr_for_counter(counter_index) {
+            Some(CSR.hpmcounter[(csr - CSR_CYCLE as u64) as usize].get_value())
+        } else {
+            None
+        }
+    }
+
+    /// Saves the internal state for PMU counters. Stops started counters, and resets all configured counters.
+    /// This should be called in anticipation of an outbound context switch.
+    pub fn save_counters(&mut self) {
+        use crate::{println, CONSOLE_DRIVER};
+        use sbi::api::pmu;
+        use PmuCounterState::*;
+        for (i, state) in self.counter_state.iter_mut().enumerate() {
+            if matches!(state, Configured(_) | Started(_)) {
+                // Unwrap ok: Max value is 64
+                let counter_index = u64::try_from(i).unwrap();
+                if let Started(c) = state {
+                    if let Some(value) = Self::get_counter_csr_value(counter_index) {
```

Comment:
It's impossible for this to fail right? The counter must have a valid CSR if we were able to start, so I would just `unwrap()`